compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchWorkgroups.cpp

```diff
@@ -0,0 +1,371 @@
+  worklist.assign(valuesDefinedAbove.begin(), valuesDefinedAbove.end());
+  while (!worklist.empty()) {
+    Value outsideValue = worklist.pop_back_val();
+    // Skip values that were already visited.
+    if (visited.count(outsideValue)) continue;
+    visited.insert(outsideValue);
+
+    Operation *definingOp = outsideValue.getDefiningOp();
+    if (!definingOp || !(isClonableIntoDispatchOp(definingOp)) ||
+        hasUnfusableUseInDispatch(outsideValue, regionOp)) {
+      valuesDefinedAbove.insert(outsideValue);
+      continue;
+    }
+    result.push_back(definingOp);
+    worklist.append(definingOp->operand_begin(), definingOp->operand_end());
+  }
+
+  return result;
+}
+
+/// Clone producers into the dispatch region.
+static LogicalResult cloneProducers(RewriterBase &rewriter,
+                                    Flow::DispatchRegionOp regionOp) {
+  SmallVector<Operation *> cloneableOps = getCloneableOps(regionOp);
+  bool sortResult = mlir::computeTopologicalSorting(cloneableOps);
+  (void)sortResult;
+  assert(sortResult && "could not compute topological sorting");
+
+  for (Operation *producer : llvm::reverse(cloneableOps))
+    if (failed(
+            clonePrecedingOpIntoDispatchRegion(rewriter, producer, regionOp)))
+      return failure();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Dispatch workgroups formation
+//===----------------------------------------------------------------------===//
+
+/// Create Flow::DispatchWorkgroupsOps
+static FailureOr<SmallVector<Flow::DispatchWorkgroupsOp>>
+createDispatchWorkgroups(mlir::TensorDimTrackingRewriter &rewriter,
+                         FunctionOpInterface funcOp,
+                         DominanceInfo const &dominanceInfo,
+                         bool generateWorkloadRegion) {
+  SmallVector<Flow::DispatchRegionOp> regionOps;
+  funcOp.walk([&](Flow::DispatchRegionOp op) { regionOps.push_back(op); });
+
+  // Clone additional producers and rewrite to DispatchWorkgroupsOp.
+  SmallVector<Flow::DispatchWorkgroupsOp> result;
+  for (auto regionOp : regionOps) {
+    if (failed(cloneProducers(rewriter, regionOp))) return failure();
+    auto maybeWorkgroupOp =
+        Flow::rewriteFlowDispatchRegionToFlowDispatchWorkgroups(regionOp,
+                                                                rewriter);
+    if (failed(maybeWorkgroupOp)) return failure();
+    result.push_back(*maybeWorkgroupOp);
+  }
+  return result;
+}
+
+/// Return `true` if the given op is contained in DispatchWorkgroupsOp or in a
+/// DispatchRegionOp.
+static bool isInDispatchRegion(Operation *op) {
+  return op->getParentOfType<Flow::DispatchWorkgroupsOp>() ||
+         op->getParentOfType<Flow::DispatchRegionOp>();
+}
+
+/// Wrap a single op in a DispatchWorkgroupsOp.
+static FailureOr<Flow::DispatchWorkgroupsOp> wrapInWorkgroupsOp(
+    mlir::TensorDimTrackingRewriter &rewriter, Operation *op,
+    bool generateWorkloadRegion) {
```

Comment:
Maybe add some more doc about the case that `generateWorkloadRegion=true`? E.g., what workload region will be generated if the option is true?